### ISSUE TITLE
wallet: don't back-date locktime when spending unconfirmed UTXOs

### ISSUE THREAD

**Issue by 0xB10C** [bitcoin/bitcoin#26527]
As mentioned in https://github.com/bitcoin/bitcoin/pull/26451#issuecomment-1313934927 example 3:

If we do anti-fee sniping, the Bitcoin Core wallet has a 10% chance to back-date the locktime of a transaction to up to 100 blocks.  Electrum has similar behavior.

https://github.com/bitcoin/bitcoin/blob/48174c0f287b19931ca110670610bd03a03eb914/src/wallet/spend.cpp#L725-L731

When spending an unconfirmed UTXO, it can happen that we back-date the locktime of our transaction to a locktime before the locktime of the unconfirmed parent. In this case, the "transaction signed earlier but only broadcast now" is unrealistic I think (please correct me if you think there is a protocol where this makes sense). This is a fingerprint that we are a Bitcoin Core or Electrum wallet.

Could be addressed together with https://github.com/bitcoin/bitcoin/issues/26526

**Comment by Tweakwild113** [comment]
> As mentioned in https://github.com/bitcoin/bitcoin/pull/26451#issuecomment-1313934927 example 3:
> 
> If we do anti-fee sniping, the Bitcoin Core wallet has a 10% chance to back-date the locktime of a transaction to up to 100 blocks.  Electrum has similar behavior.
> 
> https://github.com/bitcoin/bitcoin/blob/48174c0f287b19931ca110670610bd03a03eb914/src/wallet/spend.cpp#L725-L731
> 
> When spending an unconfirmed UTXO, it can happen that we back-date the locktime of our transaction to a locktime before the locktime of the unconfirmed parent. In this case, the "transaction signed earlier but only broadcast now" is unrealistic I think (please correct me if you think there is a protocol where this makes sense). This is a fingerprint that we are a Bitcoin Core or Electrum wallet.
> 
> related https://github.com/bitcoin/bitcoin/issues/26526
> electrum issue: https://github.com/spesmilo/electrum/issues/8073

Count me in

**Comment by Xekyo** [comment]
Good catch, it would be good to not backdate when spending unconfirmed inputs, or at least limit the locktime to the maximum of the parent transactions' locktimes. The former is probably easier.

**Comment by willcl-ark** [comment]
@danielabrozzoni you might like to take a look here and #26526?

**Comment by danielabrozzoni** [comment]
Hey! I spent some time thinking about this, ~~and implemented it in #34449~~ (I closed this one, a revised one will be opened).
Instead of not backdating at all, I decided to limit the locktime to the maximum of the parent transactions' locktimes. 

However... While writing the PR description, I realized, why would we enforce this only for unconfirmed parents? Wouldn't it make sense to make sure that the fee sniping nlocktime is >= the locktime of any parent, regardless of whether they're unconfirmed or not?

We want to use anti fee sniping while preserving some plausible deniability, so that it’s not obvious from a transaction whether we’re fee sniping or simply using nLockTime. If a child transaction has an nLockTime earlier than its parent’s nLockTime, even when the parent is confirmed, it’s not really possible to explain it as "signed earlier and broadcast later", and everyone can infer we're fee sniping. Am I missing something?

**Comment by achow101** [comment]
> Wouldn't it make sense to make sure that the fee sniping nlocktime is >= the locktime of any parent, regardless of whether they're unconfirmed or not?

That seems reasonable

**Comment by gmaxwell** [comment]
This is Not A Bug.

Sure, if no inputs are ANYONECANPAY then setting a locktime less aggressive than any input is semantically no different than setting it to the input. But it's harmless.  Kinda like people noting the the 'duration between blocks' can be negative. :P

Clamping to the inputs is non-trivial because locktimes are a union type between a lock height, a lock time, and a lockmeaningless-state due to sequences.  A PR attempting to 'fix' this issue in core (#34480) appears to have gotten it wrong due to these complexities, and also introduced a potential information leak over which inputs are wallet inputs vs non-wallet (e.g. in a payjoin or coinjoin).
